starknet_py/net/contract/contract_function.py

```diff
@@ -0,0 +1,481 @@
+
+
+@dataclass(frozen=True)
+class ContractData:
+    """
+    Basic data of a deployed contract.
+    """
+
+    address: int
+    abi: ABI
+    cairo_version: int
+
+    @cached_property
+    def parsed_abi(self) -> Union[Abi, AbiV1, AbiV2]:
+        """
+        Abi parsed into proper dataclass.
+
+        :return: Abi
+        """
+        if self.cairo_version == 1:
+            if _is_abi_v2(self.abi):
+                return AbiV2Parser(self.abi).parse()
+            return AbiV1Parser(self.abi).parse()
+        return AbiParser(self.abi).parse()
+
+    @staticmethod
+    def from_abi(address: int, abi: ABI, cairo_version: int = 0) -> ContractData:
+        """
+        Create ContractData from ABI.
+
+        :param address: Address of the deployed contract.
+        :param abi: Abi of the contract.
+        :param cairo_version: Version of the Cairo in which contract is written.
+        :return: ContractData instance.
+        """
+        return ContractData(
+            address=address,
+            abi=abi,
+            cairo_version=cairo_version,
+        )
+
+
+@add_sync_methods
+@dataclass(frozen=True)
+class InvokeResult(SentTransaction):
+    """
+    Result of the Invoke transaction.
+    """
+
+    # We ensure these are not None in __post_init__
+    contract: ContractData = None  # pyright: ignore
+    """Additional information about the Contract that made the transaction."""
+
+    invoke_transaction: Invoke = None  # pyright: ignore
+    """A InvokeTransaction instance used."""
+
+    def __post_init__(self):
+        assert self.contract is not None
+        assert self.invoke_transaction is not None
+
+
+@dataclass
+class CallToSend(Call):
+    _client: Client
+    _payload_transformer: FunctionSerializationAdapter
+
+
+@add_sync_methods
+@dataclass
+class PreparedFunctionCall(CallToSend):
+    async def call_raw(
+        self,
+        block_hash: Optional[str] = None,
+        block_number: Optional[Union[int, Tag]] = None,
+    ) -> List[int]:
+        """
+        Calls a method without translating the result into python values.
+
+        :param block_hash: Optional block hash.
+        :param block_number: Optional block number.
+        :return: list of ints.
+        """
+        return await self._client.call_contract(
+            call=self, block_hash=block_hash, block_number=block_number
+        )
+
+    async def call(
+        self,
+        block_hash: Optional[str] = None,
+        block_number: Optional[Union[int, Tag]] = None,
+    ) -> TupleDataclass:
+        """
+        Calls a method.
+
+        :param block_hash: Optional block hash.
+        :param block_number: Optional block number.
+        :return: TupleDataclass representing call result.
+        """
+        result = await self.call_raw(block_hash=block_hash, block_number=block_number)
+        return self._payload_transformer.deserialize(result)
+
+
+@add_sync_methods
+@dataclass
+class PreparedFunctionInvoke(ABC, CallToSend):
```

Comment:
A bit unsure about class name ending with a verb

Reply:
Agree that this this is not an ideal name. I would rather prefer to avoid `FunctionInvocation`, since it has a [different meaning](https://github.com/starkware-libs/starknet-specs/blob/v0.6.0/api/starknet_trace_api_openrpc.json#L300) in the RPC specification.

Reply:
Make sense, we have `Invoke` class anyway, so that'll do